File: src/mycode/js_stress_runner.py
"""js_stress_runner — Python wrapper for the Node.js instrumented caller.

Takes discovered exports from js_module_loader + scenario config, spawns
the Node.js stress runner, and returns raw harness output compatible with
engine.py's _parse_harness_output.

Part of Track B (Node.js callable harness). B3 wires this into the engine.
"""

from __future__ import annotations

import json
import logging
import os
import subprocess
from dataclasses import dataclass, field
from pathlib import Path

from mycode.js_module_loader import (
    ExportedFunction,
    ModuleDiscoveryResult,
    _check_node_available,
    _compile_typescript,
)

logger = logging.getLogger(__name__)

_RUNNER_SCRIPT = Path(__file__).parent / "js_stress_runner.js"

_RESULTS_START = "__MYCODE_RESULTS_START__"
_RESULTS_END = "__MYCODE_RESULTS_END__"

_DEFAULT_SCALE_LEVELS = [100, 1000, 10000]
_DEFAULT_STEP_TIMEOUT_MS = 60000
_DEFAULT_RUNNER_TIMEOUT = 300  # seconds (total process timeout)


# ── Result types ──


@dataclass
class StressRunResult:
    """Raw output from the Node.js stress runner.

    ``stdout`` contains the full process output (with markers).
    ``parsed`` contains the extracted JSON dict if markers were found.
    ``returncode`` is the process exit code.
    """
    stdout: str = ""
    stderr: str = ""
    returncode: int = 0
    parsed: dict | None = None
    error: str | None = None


# ── Public API ──


def run_stress(
    file_path: str,
    exports: list[ExportedFunction],
    scale_levels: list[int] | None = None,
    step_timeout_ms: int = _DEFAULT_STEP_TIMEOUT_MS,
    param_names: dict[str, list[str]] | None = None,
    node_path: str = "node",
    timeout: int = _DEFAULT_RUNNER_TIMEOUT,
    project_dir: str | None = None,
) -> StressRunResult:
    """Run the Node.js stress runner against discovered exports.

    Returns a StressRunResult whose ``stdout`` field contains the raw
    process output compatible with engine.py's _parse_harness_output.
    The ``parsed`` field contains the extracted JSON dict if available.
    """
    if not _check_node_available(node_path):
        return StressRunResult(error="Node.js is not available", returncode=-1)

    if not os.path.isfile(file_path):
        return StressRunResult(error=f"File not found: {file_path}", returncode=-1)

    if not exports:
        return StressRunResult(error="No exported functions to test", returncode=-1)

    # TypeScript: compile first
    ts_tmp_dir = None
    actual_path = file_path
    ext = os.path.splitext(file_path)[1].lower()

    if ext in (".ts", ".tsx"):
        import shutil
        compiled, ts_error = _compile_typescript(file_path, project_dir)
        if ts_error:
            return StressRunResult(error=ts_error, returncode=-1)
        actual_path = compiled
        ts_tmp_dir = os.path.dirname(compiled)

    try:
        return _spawn_runner(
            actual_path, exports, scale_levels or _DEFAULT_SCALE_LEVELS,
            step_timeout_ms, param_names or {}, node_path, timeout,
        )
    finally:
        if ts_tmp_dir:
            import shutil
            shutil.rmtree(ts_tmp_dir, ignore_errors=True)


def _spawn_runner(
    file_path: str,
    exports: list[ExportedFunction],
    scale_levels: list[int],
    step_timeout_ms: int,
    param_names: dict[str, list[str]],
    node_path: str,
    timeout: int,
) -> StressRunResult:
    """Spawn the Node.js runner and collect output."""
    task = json.dumps({
        "command": "stress",
        "file_path": os.path.abspath(file_path),
        "functions": [
            {"name": e.name, "arity": e.arity, "is_async": e.is_async}
            for e in exports
        ],
        "scale_levels": scale_levels,
        "step_timeout_ms": step_timeout_ms,
        "param_names": param_names,
    })

    try:
        result = subprocess.run(
            [node_path, str(_RUNNER_SCRIPT)],
            input=task,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired:
        return StressRunResult(
            error=f"Stress runner timed out after {timeout}s",
            returncode=-1,
        )
    except OSError as exc:
        return StressRunResult(
            error=f"Failed to spawn Node.js: {exc}",
            returncode=-1,
        )

    stdout = result.stdout
    stderr = result.stderr

    # Extract JSON between markers
    parsed = _extract_results(stdout)

    return StressRunResult(
        stdout=stdout,
        stderr=stderr,
        returncode=result.returncode,
        parsed=parsed,
    )


def _extract_results(stdout: str) -> dict | None:
    """Extract the JSON dict between result markers, or None."""
    start = stdout.find(_RESULTS_START)
    end = stdout.find(_RESULTS_END)
    if start == -1 or end == -1:
        return None

    json_str = stdout[start + len(_RESULTS_START):end].strip()
    try:
        return json.loads(json_str)
    except json.JSONDecodeError:
        return None

File: src/mycode/js_stress_runner.js
#!/usr/bin/env node
"use strict";

/**
 * js_stress_runner.js — Instrumented JS function caller with synthetic data.
 *
 * Spawned as a child process by the Python orchestrator.
 * Communication: JSON task on stdin, structured results on stdout.
 * Zero npm dependencies — Node.js built-ins only.
 *
 * Task format:
 *   {
 *     "command": "stress",
 *     "file_path": "/absolute/path/to/module.js",
 *     "functions": [{"name": "fn", "arity": 2, "is_async": false}],
 *     "scale_levels": [100, 1000, 10000],
 *     "step_timeout_ms": 60000,
 *     "param_names": {"fn": ["data", "options"]}
 *   }
 *
 * Output: __MYCODE_RESULTS_START__ JSON __MYCODE_RESULTS_END__
 */

const fs = require("fs");
const path = require("path");

// ── Module Loading (shared with js_module_loader.js) ──

function isAsyncFunction(fn) {
  if (typeof fn !== "function") return false;
  const ctor = fn.constructor;
  if (ctor && ctor.name === "AsyncFunction") return true;
  return /^async\s/.test(Function.prototype.toString.call(fn));
}

async function loadModule(filePath) {
  try {
    const resolved = require.resolve(filePath);
    return require(resolved);
  } catch (err) {
    const msg = String(err.message || err);
    const code = err.code || "";
    const isEsmError =
      code === "ERR_REQUIRE_ESM" ||
      code === "ERR_REQUIRE_ASYNC_MODULE" ||
      /Cannot use import statement/i.test(msg) ||
      /Unexpected token 'export'/i.test(msg);
    if (!isEsmError) throw err;
  }
  const fileUrl = "file://" + path.resolve(filePath);
  return await import(fileUrl);
}

// ── Synthetic Data Generation ──

const _ID_PATTERN = /^(id|user_?id|item_?id|index|idx|num|number)$/i;
const _NAME_PATTERN = /^(name|label|title|key|tag|slug|username)$/i;
const _DATA_PATTERN = /^(data|input|items|list|array|rows|records|entries|values|payload|batch|collection)$/i;
const _CONFIG_PATTERN = /^(config|options|opts|settings|params|context|env|meta)$/i;
const _CALLBACK_PATTERN = /^(callback|cb|fn|handler|onComplete|onError|next|done|resolve|reject)$/i;
const _BOOL_PATTERN = /^(flag|enabled|active|is[A-Z_]|has[A-Z_]|should|verbose|force|dry_?run|debug)$/i;
const _COUNT_PATTERN = /^(count|size|limit|num|amount|length|max|min|offset|page|per_?page|depth|width|height|timeout|retries|port)$/i;
const _PATH_PATTERN = /^(path|url|uri|file|filename|filepath|dir|directory|endpoint|route|href|src|dest|source|target)$/i;
const _STRING_PATTERN = /^(text|message|description|content|body|query|search|filter|email|password|token|code|type|format|mode|status|role|category)$/i;

function generateArg(paramName, scaleLevel) {
  if (!paramName) return generatePositionalArg(0, scaleLevel);

  if (_CALLBACK_PATTERN.test(paramName)) return function() {};
  if (_BOOL_PATTERN.test(paramName)) return true;
  if (_ID_PATTERN.test(paramName)) return scaleLevel;
  if (_COUNT_PATTERN.test(paramName)) return scaleLevel;
  if (_PATH_PATTERN.test(paramName)) return "/test/path/" + scaleLevel;
  if (_NAME_PATTERN.test(paramName)) return "test_item_" + scaleLevel;
  if (_STRING_PATTERN.test(paramName)) return "x".repeat(Math.min(scaleLevel, 10000));
  if (_CONFIG_PATTERN.test(paramName)) {
    return { debug: false, timeout: scaleLevel, retries: 3, mode: "test", batch: scaleLevel };
  }
  if (_DATA_PATTERN.test(paramName)) return generateDataArray(scaleLevel);

  // Fallback: treat unknown names as data at higher arities, string otherwise
  return generatePositionalArg(0, scaleLevel);
}

function generateDataArray(size) {
  const arr = new Array(size);
  for (let i = 0; i < size; i++) {
    arr[i] = { id: i, value: Math.random() * 1000, label: "item_" + i };
  }
  return arr;
}

function generatePositionalArg(position, scaleLevel) {
  switch (position) {
    case 0: return generateDataArray(scaleLevel);
    case 1: return { debug: false, timeout: scaleLevel, retries: 3, mode: "test", batch: scaleLevel };
    case 2: return "test_string_" + scaleLevel;
    default: return null;
  }
}

function generateArgs(funcEntry, scaleLevel, paramNames) {
  const arity = funcEntry.arity;
  if (arity === 0) return [];

  const names = paramNames || [];
  const args = [];
  for (let i = 0; i < arity; i++) {
    if (i < names.length && names[i]) {
      args.push(generateArg(names[i], scaleLevel));
    } else {
      args.push(generatePositionalArg(i, scaleLevel));
    }
  }
  return args;
}

// ── Context Error Detection ──

const _CONTEXT_ERROR_PATTERNS = [
  /ECONNREFUSED/i,
  /ENOTFOUND/i,
  /ECONNRESET/i,
  /connect ETIMEDOUT/i,
  /getaddrinfo/i,
  /is not defined/i,
  /Missing required/i,
  /API.?key/i,
  /OPENAI_API_KEY/i,
  /ANTHROPIC_API_KEY/i,
  /DATABASE_URL/i,
  /not initialized/i,
  /app.?context/i,
  /application context/i,
  /no current event loop/i,
  /working outside/i,
];

const _CONTEXT_ERROR_TYPES = new Set([
  "ReferenceError",
  "TypeError",
]);

function isContextError(err) {
  if (err == null) return false;
  const errType = (err && err.constructor) ? err.constructor.name : "Error";
  const msg = String(err.message || err);

  // Module-not-found is always context
  if (err && err.code === "MODULE_NOT_FOUND") return true;

  // Network errors are context
  if (err && (err.code === "ECONNREFUSED" || err.code === "ENOTFOUND" || err.code === "ECONNRESET")) return true;

  // Check message patterns
  for (const pattern of _CONTEXT_ERROR_PATTERNS) {
    if (pattern.test(msg)) return true;
  }

  return false;
}

// ── Instrumented Execution ──

async function invokeWithTimeout(fn, args, timeoutMs) {
  return new Promise((resolve, reject) => {
    let timer = null;
    let settled = false;

    timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        reject(new Error("Step timed out after " + timeoutMs + "ms"));
      }
    }, timeoutMs);

    try {
      const result = fn(...args);
      if (result && typeof result.then === "function") {
        result.then(
          (val) => { if (!settled) { settled = true; clearTimeout(timer); resolve(val); } },
          (err) => { if (!settled) { settled = true; clearTimeout(timer); reject(err); } }
        );
      } else {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          resolve(result);
        }
      }
    } catch (err) {
      if (!settled) {
        settled = true;
        clearTimeout(timer);
        reject(err);
      }
    }
  });
}

async function probeFunction(funcEntry, mod) {
  const fn = mod[funcEntry.name];
  if (typeof fn !== "function") {
    return { name: funcEntry.name, error: { type: "TypeError", message: funcEntry.name + " is not a function" } };
  }

  // Call with minimal safe args (not null — avoid false TypeError on null.property)
  const minArgs = [];
  for (let i = 0; i < funcEntry.arity; i++) {
    // First arg: empty array (most common data param), rest: empty object
    minArgs.push(i === 0 ? [] : {});
  }
  try {
    await invokeWithTimeout(fn, minArgs, 5000);
    return null; // probe passed
  } catch (err) {
    if (err != null && isContextError(err)) {
      const errType = (err.constructor) ? err.constructor.name : "Error";
      return {
        name: funcEntry.name,
        error: { type: errType, message: String(err.message || err).slice(0, 500) },
      };
    }
    // Non-context error (TypeError from bad args, etc.) — function is testable
    return null;
  }
}

async function measureStep(stepName, params, fn, stepTimeoutMs) {
  const errors = [];
  let capHit = "";
  const memBefore = process.memoryUsage();
  const t0 = performance.now();

  try {
    await invokeWithTimeout(fn, [], stepTimeoutMs);
  } catch (err) {
    const errType = (err != null && err.constructor) ? err.constructor.name : "Error";
    const errMsg = (err != null) ? String(err.message || err).slice(0, 500) : "Unknown error";

    if (/timed out/i.test(errMsg)) {
      capHit = "timeout";
    } else if (err instanceof RangeError && /call stack|Maximum/i.test(errMsg)) {
      capHit = "memory";
    }

    errors.push({
      type: errType,
      message: errMsg,
      traceback: (err && err.stack) ? err.stack.slice(-1000) : "",
    });
  }

  const elapsed = performance.now() - t0;
  const memAfter = process.memoryUsage();
  const peakMb = Math.max(memAfter.heapUsed, memBefore.heapUsed) / 1048576;

  return {
    step_name: stepName,
    parameters: params,
    execution_time_ms: Math.round(elapsed * 100) / 100,
    memory_peak_mb: Math.round(peakMb * 100) / 100,
    error_count: errors.length,
    errors: errors,
    resource_cap_hit: capHit,
  };
}

// ── Main Handler ──

async function handleStress(task) {
  const filePath = task.file_path;
  const functions = task.functions || [];
  const scaleLevels = task.scale_levels || [100, 1000, 10000];
  const stepTimeoutMs = task.step_timeout_ms || 60000;
  const paramNamesMap = task.param_names || {};

  const steps = [];
  const importErrors = [];
  const probeSkipped = [];

  // Load module
  let mod;
  try {
    mod = await loadModule(filePath);
  } catch (err) {
    const errType = (err && err.constructor) ? err.constructor.name : "Error";
    importErrors.push({
      type: errType,
      message: String(err.message || err).slice(0, 500),
      module: filePath,
    });
    return output({ steps, import_errors: importErrors, probe_skipped: probeSkipped });
  }

  // Unwrap ESM default if needed
  if (mod && mod.__esModule && mod.default && typeof mod.default === "object") {
    mod = Object.assign({}, mod.default, mod);
  }

  // Resolve functions to actual callables
  const resolved = [];
  for (const entry of functions) {
    const fn = mod[entry.name];
    if (typeof fn === "function") {
      resolved.push({
        name: entry.name,
        arity: entry.arity,
        is_async: entry.is_async || isAsyncFunction(fn),
        func: fn,
      });
    }
  }

  if (resolved.length === 0 && importErrors.length === 0) {
    // No callable functions found — nothing to stress
    return output({ steps, import_errors: importErrors, probe_skipped: probeSkipped });
  }

  // Probe each function
  const testable = [];
  for (const entry of resolved) {
    const probeResult = await probeFunction(entry, mod);
    if (probeResult) {
      probeSkipped.push(probeResult);
    } else {
      testable.push(entry);
    }
  }

  if (testable.length === 0) {
    return output({ steps, import_errors: importErrors, probe_skipped: probeSkipped });
  }

  // Run stress steps at each scale level
  const totalStartMs = performance.now();
  for (const level of scaleLevels) {
    // Time budget: skip remaining levels if 80% of total budget used
    const elapsedTotal = performance.now() - totalStartMs;
    const totalBudget = stepTimeoutMs * scaleLevels.length;
    if (elapsedTotal > totalBudget * 0.8 && steps.length > 0) {
      steps.push({
        step_name: "scale_" + level,
        parameters: { scale_level: level, functions_called: 0 },
        execution_time_ms: 0,
        memory_peak_mb: 0,
        error_count: 1,
        errors: [{ type: "Skipped", message: "Time budget exceeded (80%)", traceback: "" }],
        resource_cap_hit: "timeout",
      });
      continue;
    }

    const step = await measureStep(
      "scale_" + level,
      { scale_level: level, functions_called: testable.length },
      async () => {
        for (const entry of testable) {
          const args = generateArgs(entry, level, paramNamesMap[entry.name]);
          const fn = entry.func;
          const result = fn(...args);
          if (result && typeof result.then === "function") {
            await result;
          }
        }
      },
      stepTimeoutMs,
    );
    steps.push(step);

    // Stop escalating if step hit resource cap
    if (step.resource_cap_hit === "memory" || step.resource_cap_hit === "timeout") {
      break;
    }
  }

  output({ steps, import_errors: importErrors, probe_skipped: probeSkipped });
}

function output(data) {
  console.log("__MYCODE_RESULTS_START__");
  console.log(JSON.stringify(data));
  console.log("__MYCODE_RESULTS_END__");
}

// ── stdin reader ──

let input = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => { input += chunk; });
process.stdin.on("end", async () => {
  let task;
  try {
    task = JSON.parse(input);
  } catch (err) {
    console.error("Invalid JSON on stdin");
    process.exitCode = 1;
    return;
  }

  if (task.command === "stress") {
    try {
      await handleStress(task);
    } catch (err) {
      console.error("Fatal error:", err);
      output({ steps: [], import_errors: [], probe_skipped: [] });
      process.exitCode = 1;
    }
  } else {
    console.error("Unknown command: " + task.command);
    process.exitCode = 1;
  }
});
